### ISSUE TITLE
no-useless-return causes a stack overflow with some unreachable loops

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#7583]
**Tell us about your environment**

* **ESLint Version:** master
* **Node Version:** 7.1.0
* **npm Version:** 3.10.9

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

(none)

**What did you do? Please include the actual source code causing the issue.**

```js
/* eslint no-useless-return: error */

function foo() {
  return;
  while (foo) return;
  foo;
}
```

**What did you expect to happen?**

ESLint should have either reported a useless return statement, or exited without reporting any errors.

**What actually happened? Please include the actual, raw output from ESLint.**

`no-useless-return` threw an error:

```
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at WeakMap.get (native)
    at isReturned (/path/to/eslint/lib/rules/no-useless-return.js:119:41)
    at Array.filter (native)
    at markReturnStatementsOnSegmentAsUsed (/path/to/eslint/lib/rules/no-useless-return.js:179:22)
    at Array.forEach (native)
    at markReturnStatementsOnSegmentAsUsed (/path/to/eslint/lib/rules/no-useless-return.js:180:22)
    at Array.forEach (native)
    at markReturnStatementsOnSegmentAsUsed (/path/to/eslint/lib/rules/no-useless-return.js:180:22)
    at Array.forEach (native)
    at markReturnStatementsOnSegmentAsUsed (/path/to/eslint/lib/rules/no-useless-return.js:180:22)

```
